examples/usage.rs

```diff
@@ -0,0 +1,70 @@
+#![feature(try_blocks)]
+// Overrides the std library's `Result` type.
+
+use core::fmt;
+use std::fs::File;
+use std::io;
+
+#[derive(Debug)]
+enum MyError {
+    Unlucky,
+    Io(io::Error),
+    TooSmall(u64),
+}
+
+impl fmt::Display for MyError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Unlucky => write!(f, "Not this time!"),
+            Self::Io(_) => write!(f, "I/O error"),
+            Self::TooSmall(size) => write!(f, "File too small: {} bytes", size),
+        }
+    }
+}
+
+impl std::error::Error for MyError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            MyError::Unlucky => None,
+            MyError::Io(e) => e.source(),
+            MyError::TooSmall(_) => None,
+        }
+    }
+}
+
+impl From<io::Error> for MyError {
+    fn from(e: io::Error) -> Self {
+        Self::Io(e)
+    }
+}
+
+fn file_size(path: &str) -> propagate::Result<u64, MyError> {
+    try {
+        // The `?` operator coerces `std::result::Result<_, io::Error>` into `Result<_, MyError>`.
+        let size = File::open(path)?.metadata()?.len();
+
+        if size < 1024 {
+            Err(MyError::TooSmall(size))?
+        }
+
+        size
+    }
+}
+
+fn maybe_file_size(path: &str) -> propagate::Result<u64, MyError> {
+    let lucky = (path.len() % 2) == 0;
+
+    try {
+        if !lucky {
+            Err(MyError::Unlucky)?
+        }
+
+        file_size(path)?
+    }
+}
+
+fn main() -> propagate::Result<(), MyError> {
+    let size = maybe_file_size("foo.txt")?;
+    println!("File size: {} KiB", size / 1024);
+    propagate::Ok(())
+}
```

Comment:
I tried experimenting with using the `Termination` trait here to handle printing out errors. The end result looks like this

<pre><font color="#A1B56C"><b>❯</b></font> <font color="#A1B56C">cargo</font><font color="#D8D8D8"> </font><font color="#A1B56C">run</font><font color="#D8D8D8"> </font><font color="#A1B56C">--example</font><font color="#D8D8D8"> </font><font color="#A1B56C">usage</font>
<font color="#F7CA88"><b>warning</b></font><b>:</b> unused manifest key: package.author
<font color="#A1B56C"><b>   Compiling</b></font> propagate v0.0.1 (/home/jlusby/git/yaahc/propagate)
<font color="#A1B56C"><b>    Finished</b></font> dev [unoptimized + debuginfo] target(s) in 0.48s
<font color="#A1B56C"><b>     Running</b></font> `target/debug/examples/usage`
Error: Not this time!

Return Trace: 
   0: examples/usage.rs:59
   1: examples/usage.rs:67
</pre>

Reply:
Neat!